### ISSUE TITLE
TypeGuard incompatible with generic TypeVar

### ISSUE THREAD

**Issue by tekumara** [microsoft/pyright#4058]
```python
from typing import Any, List, Type, TypeVar
from typing_extensions import TypeGuard, reveal_type

T = TypeVar('T', str, float)


def istype(t: Type[Any], ty: Type[T]) -> TypeGuard[Type[T]]:
    return t == ty

def narrow(s: Type[T]) -> List[T]:
    if istype(s, str):
        reveal_type(s)
        return ["s"]
    return []
```

```
$ pyright
pyright 1.1.275
/tmp/typeg.py
  /tmp/typeg.py:12:21 - information: Type of "s" is "Type[str]"
  /tmp/typeg.py:13:16 - error: Expression of type "list[str]" cannot be assigned to return type "List[T@narrow]"
    "list[str]" is incompatible with "List[T@narrow]"
      TypeVar "_T@list" is invariant
        Type "str" cannot be assigned to type "T@narrow" (reportGeneralTypeIssues)
```

**Comment by hmc-cs-mdrissi** [comment]
This is a typeguard vs strict type guard issue. Strict preserves the typevar information, while non strict loses it. https://github.com/python/typing/discussions/1013

**Comment by erictraut** [comment]
This is a case I hadn't considered previously.

After thinking about it more, I don't see a good fix. A type checker has no knowledge about what logic a `TypeGuard` function performs. The semantics are unknown (and generally unknowable). By returning a `True` value, the user`TypeGuard` function attests that the return type is as specified in the return type, but that's all the type checker can assume.

In your example, the return type evaluates to `TypeGuard[str]` after solving the TypeVar T@istype, so the narrowed type of `s` becomes `str`. To avoid the error, the narrowed type would need to become a [conditional type](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#constrained-type-variables-and-conditional-types) `str*`, effectively an intersection between `str` and `T@narrow`.

For built-in type guards such as `isinstance` and `isclass` calls, the narrowing logic has enough knowledge of the semantics of the call, so it can safely create a conditional type and make it conditioned on a particular constrained type. For user-defined type guards, there's not enough information for the logic to make that association. It can't assume that the first argument passed to the `TypeGuard` function has relationship to the second or subsequent arguments.

@hmc-cs-mdrissi, I don't think problem gets any better with `StrictTypeGuard`. The same issue applies there.

I think you'll need to explore alternative solutions. I don't know enough about the real problem you're trying to solve here, so I can't offer any concrete suggestions.

**Comment by tekumara** [comment]
Thank you for the detailed explanation. I can probably work around this using overloads instead.
